e2e/tests/node_not_ready/node_power_off.robot

```diff
@@ -0,0 +1,320 @@
+    When Power off the node ${node_number_without_replica}[0]
+    And Sleep ${POD_EVICTION_TIME_OUT} seconds 
+    Then Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Node ${node_number_without_replica}[0] state should be NotReady
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Volume state should be healthy
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Engine state should be running
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Replica on the node ${node_number_with_replica}[0] state should be running
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Replica on the node ${node_number_with_replica}[1] state should be running
+    When Power on the node ${node_number_without_replica}[0]
+    Then Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Node ${node_number_without_replica}[0] state should be Ready
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Volume state should be healthy
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds     ${retry_interval} seconds    Engine state should be running
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Replica on the node ${node_number_with_replica}[0] state should be running
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Replica on the node ${node_number_with_replica}[1] state should be running
+    [Teardown]    cleanup_resources
+
+14-Power off 2 replica RWO no attach node with replica time out
+    [Documentation]
+    [Tags]    NodePoweringOff    RWO    2Replicas    AttachReplicaDiffNode    AbnormalReplicaNode    AbnormalAttachedNode
+    Given Create ${volume_size_gb} GB RWO volume with 2 replicas
+    And Get node index with replica
+    And Attach volume to the node ${node_number_with_replica}[0]
+    And Write data into mount point
+    When Power off the node ${node_number_with_replica}[1]
+    And Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Sleep ${POD_EVICTION_TIME_OUT} seconds
+    Then Run keyword And Continue On Failure    Wait Until Keyword Succeeds    ${retry_time_out} seconds    ${retry_interval} seconds    Node ${node_number_with_replica}[1] state should be NotReady
```

Comment:
This kind of test definition is less human-readable than our existing test skeleton defined in `tests/test_*.py`.

Our existing test skeleton defined in `tests/test_*.py` is 100% human readable, anyone who wants to add a test skeleton just needs to describe what the test case wants to do step by step, and doesn't need any prior knowledge.  

But for the test definitions in these `*.robot` files, there are lots of variable references and string interpolations in the test definition like `${node_number_with_replica}[0]`. Also there are built-in keywords like `Run keyword And Continue On Failure` or `Wait Until Keyword Succeeds` directly being used.

Which means:
(1) Extra efforts for the user wants to contribute to our test code base (learn variable naming/referencing and robot framework built-in functions)
(2) More difficult to understand the test case than before because it's robot framework syntax instead of simple sentence.

I'm worried about these will reduce the willingness of community users (or even the developers) to contribute to our test code.

I think it's possible that we can achieve higher abstraction and better decoupling to let people write test definition in the simplest way and don't even have to deal with test framework specific things.

WDYT @innobead @longhorn/qa

Reply:
I think the simplest way to do it is just copy and paste test steps from https://confluence.suse.com/pages/viewpage.action?spaceKey=LON&title=Negative+Testing or any manual test case to the `*.robot` file, and hide all the implementation details in custom keywords. Is it possible?

Reply:
When using any framework, it definitely requires some costs but u also will gain the benefits, even pytest. If don't know it well, it would cause potential redundancy, unstructured design, etc.

For any solutions, they definitely have pros and cons. How to attract community to get involved in contributing primarily relies on the development and contribution guide which is what we should improve.

Let's focus on the review itself. The thing is just started and we should think how to make it better to resolve the points you mentioned above.

Reply:
> I think the simplest way to do it is just copy and paste test steps from https://confluence.suse.com/pages/viewpage.action?spaceKey=LON&title=Negative+Testing or any manual test case to the `*.robot` file, and hide all the implementation details in custom keywords. Is it possible?

Anything is possible, but probably u need to craft that a bit. Think in details.

Reply:
BTW, everyone can check @c3y1huang  e2e test cases which are always BDD for the test skeletons she implemented.

Reply:
This is an example. Just for reference.
https://github.com/longhorn/longhorn-tests/blob/f8e716273ab243ea07b7c51e646566b5b297e740/e2e/tests/replica_resiliency_test.robot#L15

Reply:
Regarding making test case step descriptions more readable, I have the first executable version on another branch and will continue to modify it.

```robotframework
*** Test Cases ***    
1-RWO 3 replicas abnormal attached and replica node 
    [Tags]    RestartKubelet    RWO    3Replicas    AttachHasReplica    AttachReplicaSameNode    AbnormalAttachedNode    AbnormalReplicaNode
    # Recover: After restart the node, volume should be detached and replica is failed. Need manual to do reattached.
    Given Create ${volume_size_gb} GB RWO volume with 3 replicas
    And Attach volume to the node 1 with replica node
    And Write data into mount point
    When Restart Kubelet the node 1 with replica node
    Then Node 1 with replica node state should be Ready
    And Volume state should be healthy
    And Engine state should be running
    And Replica on the node 0 state should be running
    And Replica on the node 1 state should be running
    And Replica on the node 2 state should be running
    And Data is intact
    [Teardown]    cleanup_resources
```
![image](https://github.com/longhorn/longhorn-tests/assets/17548901/ca37b129-f11e-43d3-96c9-71d3c9366262)

Reply:
In current version, I try to make the content more human readable and it will be like below.
https://github.com/longhorn/longhorn-tests/blob/15ca9dcba0534243fb68b9d0060173eafd129a5f/e2e/tests/node_not_ready/node_power_off.robot#L26-L44